components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/OptimizedApplicationConfig.java

```diff
@@ -248,37 +293,57 @@ private LocalAuthenticatorConfig[] getLocalAuthenticatorConfigs(List<String> loc
             try {
-                IdentityProvider idp = manager.getIdPByResourceId(federatedIdPResourceIds.get(i), tenantDomain,
-                        false);
-                if (idp == null) {
+                IdentityProvider idPByResourceId = manager.getIdPByResourceId(optimizedFederatedIdP.getIdpResourceId(),
+                        tenantDomain, false);
+                if (idPByResourceId == null) {
                     throw new SessionDataStorageOptimizationClientException(
                             String.format("Cannot find the IdP by the resource Id: %s Tenant Domain: %s",
-                                    federatedIdPResourceIds.get(i), tenantDomain));
+                                    optimizedFederatedIdP.getIdpResourceId(), tenantDomain));
+                }
+                if (StringUtils.equals(idPByResourceId.getDefaultAuthenticatorConfig().getName(),
+                        optimizedFederatedIdP.getDefaultAuthenticatorName())) {
+                    idPList.add(idPByResourceId);
+                } else {
+                    /**
+                     * For applications that use custom connector IdPs as a sign-in method, admin can select and use one
+                     * of the authenticators available in the custom connector IDP, other than the default authenticator
+                     * of the IDP. So need to find the exact authenticator configured in the SP.
+                     */
+                    IdentityProvider clonedIdP = createIdPClone(idPByResourceId);
+                    for (FederatedAuthenticatorConfig fedAuthConfig :
+                            idPByResourceId.getFederatedAuthenticatorConfigs()) {
+                        if (StringUtils.equals(fedAuthConfig.getName(),
+                                optimizedFederatedIdP.getDefaultAuthenticatorName())) {
+                            clonedIdP.setDefaultAuthenticatorConfig(fedAuthConfig);
+                            break;
+                        }
+                    }
+                    idPList.add(clonedIdP);
                 }
-                idPs[i] = idp;
-            } catch (IdentityProviderManagementClientException e) {
-                throw new SessionDataStorageOptimizationClientException(
-                        String.format("IDP management client error. Failed to get the IdP by the name: %s " +
-                                        "Tenant Domain: %s", federatedIdPResourceIds.get(i), tenantDomain), e);
-            } catch (IdentityProviderManagementServerException e) {
-                throw new SessionDataStorageOptimizationServerException(
-                        String.format("IDP management server error. Failed to get the IdP by the name: %s " +
-                                        "Tenant Domain: %s", federatedIdPResourceIds.get(i), tenantDomain), e);
             } catch (IdentityProviderManagementException e) {
```

Comment:
IdentityProviderManagementClientException is caught separately here to handle the client-side errors and return Client Exception

Reply:
addressed with Addressed with https://github.com/wso2/carbon-identity-framework/commit/87aac03e1fe4a099ff0d87cfced2667a1a3f74ad